XFormInstance attachments -> blobs

millerdev opened the pull request:
This has been on staging for a while. I'll run the migration once this is merged and applications and multimedia have been migrated.

@dannyroberts @snopoke cc @calellowitz

millerdev committed:
XFormInstance attachments -> blobs

millerdev committed:
Adapt uses of _attachments to support BlobMixin

millerdev committed:
Refactor DocumentTransform to support blob db

millerdev committed:
Refactor clone_domain command to support blob db

millerdev committed:
Fix custom/legacy attachment access

millerdev committed:
Remove delete_jadjars management command

Command uses couchdb.put_attachments(doc, ...), which is not compatible with the blob db.

millerdev committed:
Add XFormInstance blob db migration

dimagimon commented on corehq/apps/domainsync/config.py:
```diff
@@ -11,29 +12,42 @@ def __init__(self, doc, database, exclude_attachments=False):
+    def save():
+        try:
+            database.save_doc(transform.doc, force_update=True)
+        except ResourceNotFound, e:
```
F841 local variable 'e' is assigned to but never used

dimagimon commented on custom/_legacy/pact/api.py:
```diff
@@ -147,7 +148,7 @@ def return_iterator():
 #                    continue
                 try:
-                    xml_str = db.fetch_attachment(data_row['_id'], 'form.xml').replace("<?xml version=\'1.0\' ?>", '').replace("<?xml version='1.0' encoding='UTF-8' ?>", '')
+                    xml_str = BlobHelper(data_row, db).fetch_attachment('form.xml').replace("<?xml version=\'1.0\' ?>", '').replace("<?xml version='1.0' encoding='UTF-8' ?>", '')
```
E501 line too long (178 > 115 characters)

dimagimon commented on corehq/apps/domainsync/config.py:
```diff
@@ -11,29 +12,42 @@ def __init__(self, doc, database, exclude_attachments=False):
+            transform.doc['_rev'] = rev
+            database.save_doc(transform.doc)
+    if transform.attachments:
+        obj = BlobHelper(doc, database)
```
F821 undefined name 'doc'

millerdev commented on corehq/couchapps/attachments/views/attachments/map.js:
```diff
@@ -10,6 +10,13 @@ function (doc) {
+                    media += 1;
+                    attachments[key] = doc.external_blobs[key];
+                }
+            }
```
I assume this is necessary to keep the view functioning properly, but I haven't made many changes like this before. It seems a shame to have kick off a huge reindex for this since when it goes live the index will be up to date and only new `XFormInstance` docs created after the merge will need this new index code. Is there a way to avoid the reindex in this type of situation?

snopoke replied:
Unfortunately I don't think there is a way to get around the reindex. There may be a way to decouple it from the rest of the changes though so that you can get it done prior to merging the rest. @dannyroberts any thoughts

dannyroberts replied:
You may want to just kick this reindex off Friday after the deploy so it can run the weekend. I believe it won't last longer than the weekend.

millerdev replied:
Because of other issues that have come up on this PR, I'm holding off on starting the preindex until I have resolved them.

millerdev committed:
lint and bugfix

snopoke commented on corehq/form_processor/backends/couch/update_strategy.py:
```diff
@@ -315,20 +315,18 @@ def _apply_attachments_action(self, attachment_action, xform=None):
-                return base64.b64decode(xform._attachments[name]['data'])
-            else:
-                return XFormInstance.get_db().fetch_attachment(attachment_action.xform_id, name)
+        def fetch_attachment(name, _form=[]):
```
Clever though confusing use of a mutable argument.

Also if `xform` is not None couldn't you just use that to get the attachment?

millerdev replied:
yeah, it's a tradeoff between adding another variable outside of this function or doing this hacky thing with a (private, indicated by leading underscore) parameter with default value. The amount of code you need to read to understand how it's being used is only 4 lines, so seemed like a reasonable tradeoff to me. I'll change it if you think it's too clever.

Good point on using `xform`, I'll fix that part.

snopoke commented on corehq/pillows/mappings/xform_mapping.py:
```diff
@@ -47,6 +47,10 @@
+        'external_blobs': {
+            'dynamic': False,
+            'type': 'object'
+        },
```
another ES reindex, fun times

snopoke replied:
don't you need to update the index name?

millerdev replied:
That's a good question. I don't know. Can you advise on what needs to be done?

emord replied:
@millerdev you have to change this line: https://github.com/dimagi/commcare-hq/blob/3f49c8a5b711d5434dae72f1e26628797e5a76a4/corehq/pillows/mappings/xform_mapping.py#L6

there's another line in the pillowtop tests that needs to be changed as well (should be able to search for the string)

millerdev committed:
Use provided xform to fetch attachments

millerdev committed:
Fix broken ES accessors test

millerdev commented on corehq/couchapps/attachments/views/attachments/map.js:
```diff
@@ -10,6 +10,13 @@ function (doc) {
                     attachments[key] = doc._attachments[key];
                 }
             }
+            for (var key in doc.external_blobs) {
```
Seems like this will throw an exception if `doc.external_blobs` is undefined, and same above for `doc._attachments`... I think that means I need to fix this and reindex staging again. Can someone sanity check me on that, please?

millerdev replied:
Hmm, maybe not?
```js
$ node
> doc = {}
{}
> for (var k in doc.blobs) { console.log("k=" + k); }
undefined
> doc.blobs
undefined
```

millerdev commented on corehq/ex-submodules/couchforms/models.py:
```diff
@@ -96,9 +96,11 @@ class XFormOperation(DocumentSchema):
 
-class XFormInstance(SafeSaveDocument, UnicodeMixIn, ComputedDocumentMixin,
-                    CouchDocLockableMixIn, AbstractXFormInstance):
+class XFormInstance(DeferredBlobMixin, SafeSaveDocument, UnicodeMixIn,
```
`DeferredBlobMixin` seemed like a cool idea because it allows attachments added to a form to be persisted in the blob db when the form is `save()`d. However, it also seems to be the source of [this test failure](https://travis-ci.org/dimagi/commcare-hq/jobs/140928493).

Here's my current hypothesis on the issue:
`XFormInstance`s are sometimes [saved with `bulk_save`](https://github.com/dimagi/commcare-hq/blob/3f49c8a5b711d5434dae72f1e26628797e5a76a4/corehq/form_processor/backends/couch/processor.py#L66), which means the [`save` method](https://github.com/dimagi/commcare-hq/blob/3f49c8a5b711d5434dae72f1e26628797e5a76a4/corehq/blobs/mixin.py#L412) is not called, and forms' attachments are not saved. The reason that this previously worked is that the attachments were encoded in the doc dict that was uploaded to couch either with `xforminstance.save()` or `XFormInstance.get_db().bulk_save(docs)`. Note: I found this issue while analyzing the test failure; attachments not being saved may not be the direct cause of the test failure.

@dannyroberts / @snopoke: my current plan is to build some `atomic_blobs`-like thing that works with `bulk_save` and then either override `bulk_save` or track down all the uses of `bulk_save`-like things that are currently saving `DeferredBlobMixin` objects and wrap them with attachment saving logic. The override seems easy enough but expands the code that needs to know about the blob db. The second seems hard and possibly error prone (easy to miss one). Do you have any other ideas of how I could address this issue?

Aside: the use of `bulk_save` in our codebase seems to imply an argument against putting any custom logic in the `Document.save` method because it means it is possible to save an object without executing its custom save logic.

snopoke replied:
The requirement for this comes from the form processing where we want the processing to be atomic for each form submission. If we can approach that problem differently then I think we can get rid of the need for deferred attachments.

The only reason to use `Class.get_db().bulk_save(docs)` is if you're saving docs that have a different `doc_type` from the `Class` (or in generic doc processing code e.g. `IterDB`) . In our code this mostly happens when we do a soft delete (where we append `-deleted` to the `doc_type`). The other place it happens is for `XFormInstance` objects which can have a range of `doc_types`: https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/couchforms/models.py#L50-L50

millerdev committed:
Fix XFormInstance blobs not being saved

millerdev commented on corehq/blobs/migrate.py:
```diff
@@ -240,6 +241,15 @@ def migrate(self, *args, **kw):
+        xform.XFormDeprecated,
+        xform.XFormDuplicate,
+        xform.XFormError,
+        xform.SubmissionErrorLog,
```
Note to self: looks like [`HQSubmission`](https://github.com/dimagi/commcare-hq/pull/12333/files#diff-93c384a6794b8c149c3fe4243e26957aR29) is missing from this list. Not sure if that's important.

millerdev replied:
@dannyroberts: after searching in the code, it looks like `HQSubmission` may be a historical artifact. Is it important to migrate their attachments to the blob db?

dannyroberts replied:
Good question. I'm not 100% sure, but I'd default to also migrating those attachments. Might be worth checking all_docs_by_doc_type to see how many of these exist?

millerdev committed:
Adapt to various attachment interfaces

Fixes corehq.apps.reports.tests.test_form_export:FormMultimediaExportTest.test_extract_form_attachment_info

millerdev committed:
Adapt form submission test to blob db

Fixes corehq.apps.receiverwrapper.tests.test_submissions:SubmissionTest.test_submit_bare_form

millerdev committed:
Fix bulk_atomic_blobs

millerdev committed:
Fix missing attachments after edit form

Fixes corehq.apps.commtrack.tests.test_rebuild:RebuildStockStateTest.test_edit_submissions_simple

millerdev committed:
Fix loss of attachment content type

millerdev committed:
typo

Fixes corehq.apps.receiverwrapper.tests.test_submit_errors:SubmissionErrorTest.testSubmitBadXML

millerdev committed:
Copy deferred blobs on dedupe xform

Fixes corehq.apps.receiverwrapper.tests.test_submit_errors:SubmissionErrorTest.testSubmitDuplicate

millerdev committed:
Allow curly braces in blob filenames

For corehq.apps.tzmigration.tests.test_timezonemigration:TimeZoneMigrationTest.test_migration

millerdev committed:
Update test data for blob db

Fixes corehq.apps.tzmigration.tests.test_timezonemigration

millerdev committed:
Erroneous tuple

Fixes corehq/ex-submodules/casexml/apps/case/tests/test_bugs.py:CaseBugTest.testDateInCasePropertyBug

millerdev committed:
Improve test cleanup for REUSE_DB

millerdev committed:
Fix couchforms.tests.test_post

millerdev committed:
Bump ES index versions

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/xform-blobs

Conflicts:
	corehq/pillows/mappings/reportxform_mapping.py

dimagimon commented on corehq/blobs/mixin.py:
```diff
@@ -420,6 +420,31 @@ def save(self):
+    :param docs: A list of model objects.
+    """
+    # TODO test with all non-atomic_blobs docs
+    save = lambda:None
```
E231 missing whitespace after ':'

millerdev committed:
lint

millerdev committed:
Update index names in tests

millerdev committed:
missed one

dannyroberts commented:
(looks good to me when ready!)

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/xform-blobs

Conflicts:
	testapps/test_pillowtop/tests/data/all-pillow-meta.json

millerdev committed:
Add HQSubmission to xform blob migration

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/xform-blobs

Conflicts:
	testapps/test_pillowtop/tests/data/all-pillow-meta.json

czue commented:
@millerdev the xforms index rename causes a huge ES reindex that we shouldn't do mid-week. reverted in https://github.com/dimagi/commcare-hq/pull/12539. we should probably redo this over the weekend.

millerdev commented:
@czue I discussed this with @dannyroberts. I thought we had deployed the reindex previously, so wasn't aware that that was going to cause issues. Did this get deployed? If so, it seems dangerous to revert this branch because form submissions that are in blob db could be lost (although maybe only temporarily until we get this merged again).

dannyroberts commented:
When we talked and I merged, I was only thinking about the CouchDB reindex; I forgot that there was also an ES reindex involved (due to the changes to corehq/pillows/mappings/xform_mapping.py). @millerdev, as far as I know, only the CouchDB reindex was separately PR'd, merged, run, and deployed; the ES reindex is still in this PR as @czue pointed out. Does that match up with your model of what's happened?

snopoke commented:
Yes, that's correct. This PR has changes to both the `xforms` index and `report_xforms` which will both need to be reindexed. This PR never got deployed (Cory cancelled the deploy when the reindex started).

millerdev commented:
ok, that all seems right to me. Feels pretty bad to have botched the roll-out like that, but I'm glad we caught it before the deploy was fully rolled out. We'll try again on the weekend.

Would it makes sense to separate out the ES reindex like we did for couch and deploy that separately, or is it safe/reasonable to keep them combined?

dannyroberts commented:
I'd say it's certainly ok to split out the reindex as a separate PR (since it's only adding properties), but also probably ok to merge them together at this point since the rest of the PR is ready to go as well.

It also occurs to me that the property being added to the mapping currently doesn't exist for any xforms, and so the reindex is a bit of a waste. Still, it may be operationally the easiest thing to do.

millerdev commented:
> It also occurs to me that the property being added to the mapping currently doesn't exist for any xforms, and so the reindex is a bit of a waste.

Does this mean it would be OK to roll out the code without bumping the index versions? Or would that cause ES to ignore the changes in the index schema? I think what I'm asking is: is there a way to update the index schema/code without actually re-indexing existing data?

benrudolph commented:
I think all forms _will_ have the external blobs property and a reindex will be necessary. when the migration happens, the form gets saved with that new property and gets processed by the pillow to send it into ES again

czue commented:
yeah unfortunately i don't think we currently have any reliable workflow for changing an index that doesn't involve reindexing the data.

snopoke commented:
If we're just adding a new field for which existing data won't have a value
then we should just be able to update the mapping via the put mapping api
<https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html>
which
we could do with a management command. This will make any new data
searchable on those new fields but any old data will not be searchable.
Here's a good blog on changing mappings:
https://www.elastic.co/blog/changing-mapping-with-zero-downtime

millerdev commented:
I think I'd rather err on the side of caution and not introduce a new reindexing workflow on this PR. It can wait until the weekend.

millerdev commented:
looks like we forgot to re-merge this over the weekend. Will set a reminder to merge it this coming friday
